Python/Product/Analysis/Interpreter/Ast/AstAnalysisWalker.cs

```diff
@@ -0,0 +1,260 @@
+            if (expr is LambdaExpression) {
+                return _interpreter.GetBuiltinType(BuiltinTypeId.Function);
+            }
+
+            return null;
+        }
+
+        private IMember GetInScope(string name) {
+            foreach (var m in _scope) {
+                if (m == null) {
+                    continue;
+                }
+                IMember obj;
+                if (m.TryGetValue("__class__", out obj)) {
+                    return obj;
+                }
+            }
+            return null;
+        }
+
+        private IPythonType CurrentClass {
+            get {
+                var m = _scope.Peek();
+                if (m != null) {
+                    IMember cls;
+                    m.TryGetValue("__class__", out cls);
+                    return cls as IPythonType;
+                }
+                return null;
+            }
+        }
+
+        public override bool Walk(AssignmentStatement node) {
+            var m = _scope.Peek();
+            if (m != null) {
+                var value = GetValueFromExpression(node.Right, m);
+                foreach (var ne in node.Left.OfType<NameExpression>()) {
+                    m[ne.Name] = value;
+                }
+            }
+            return base.Walk(node);
+        }
+
+        public override bool Walk(ImportStatement node) {
+            var m = _scope.Peek();
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i].Names[0];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(node.AsNames[i])
+                            );
+                        }
+                    }
+                } catch (IndexOutOfRangeException) {
+                }
+            }
+            return false;
+        }
+
+        public override bool Walk(FromImportStatement node) {
+            var m = _scope.Peek();
+            if (node.Root.MakeString() == "__future__") {
+                return false;
+            }
+
+            if (m != null && node.Names != null) {
+                try {
+                    for (int i = 0; i < node.Names.Count; ++i) {
+                        var n = node.AsNames?[i] ?? node.Names[i];
+                        if (n != null) {
+                            m[n.Name] = new AstPythonConstant(
+                                _interpreter.GetBuiltinType(BuiltinTypeId.Module),
+                                GetLoc(n)
+                            );
+                        }
+                    }
+                } catch (IndexOutOfRangeException) {
+                }
+            }
+            return false;
+        }
+
+        public override bool Walk(FunctionDefinition node) {
```

Comment:
>public override bool Walk(FunctionDefinition node) { [](start = 8, length = 52)

If you did 2 passes here you'd get significantly better results as you could collect the functions / classes first and then get any forward references, e.g:

def f():
    return g

def g():
    ...